File: Year2023/src/Day5.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import java.util.PrimitiveIterator;
import java.util.stream.LongStream;

/**
 * <a href="https://adventofcode.com/2023/day/5">Day 5</a>
 */
public class Day5 extends Day<Long> {
	@Override
	public Long run1(List<String> a) {
		List<List<String>> input = getParsedInput();

		long best = Integer.MAX_VALUE;
		for(Long l : getSeeds(input)) {
			for(Map map : getMaps(input)) {
				l = map.getValue(l);
			}
			if(l < best) {
				best = l;
			}
		}
		return best;
	}

	@Override
	public Long run2(List<String> a) {
		//noinspection ConstantConditions
		List<List<String>> input = getParsedInput();
		Map[] maps = getMaps(input);

		List<Long> seeds = getSeeds(input);
		int size = seeds.size();
		long[][] ranges = new long[size / 2][];
		for (int i = 0; i < size; i += 2) {
			long low = seeds.get(i);
			long high = seeds.get(i + 1);
			ranges[i / 2] = new long[]{low, high};
		}

		Utils.reverse(maps); // reverse the maps so we can go backwards

		// get the largest value from the last map
		PrimitiveIterator.OfLong itr = maps[0].getMaxRange();

		// send them all backwards to find their original values
		while(itr.hasNext()) {
			long next = itr.nextLong();
			long changed = next;

			for(Map map : maps) {
				changed = map.getKey(changed);
			}

			// if any of the original values are in the ranges, we found the answer
			for(long[] range : ranges) {
				long low = range[0];
				long high = low + range[1];
				if(low <= changed && changed < high) {
					return next;
				}
			}
		}
		throw new AssertionError("No solution found");
	}

	private List<List<String>> getParsedInput() {
		List<List<String>> parsedInput = new ArrayList<>();
		List<String> current = new ArrayList<>();
		for(String line : Main.getInput(5)) {
			if(line.isEmpty()) {
				parsedInput.add(current);
				current = new ArrayList<>();
			} else {
				current.add(line);
			}
		}
		if(!current.isEmpty())
			parsedInput.add(current);
		return parsedInput;
	}

	private List<Long> getSeeds(List<List<String>> input) {
		String[] seeds = Utils.SPACES.split(input.get(0).get(0).substring(7));
		List<Long> list = new ArrayList<>();
		for(String seed : seeds) {
			list.add(Utils.fastParseLong(seed));
		}
		return list;
	}

	private Map[] getMaps(List<List<String>> input) {
		input = new ArrayList<>(input);
		input.remove(0);
		Map[] maps = new Map[input.size()];
		for(int i = 0; i < input.size(); i++) {
			maps[i] = Map.create(input.get(i));
		}
		return maps;
	}

	record Map(Entry[] entries) {
		static Map create(List<String> lines) {
			Entry[] entries = new Entry[lines.size() - 1];
			for(int i = 1; i < lines.size(); i++) {
				entries[i - 1] = Entry.create(lines.get(i));
			}
			// sort so we can binary search later
			Arrays.sort(entries, Comparator.comparingLong(o -> o.destination)); // faster than Entry::destination for some reason
			return new Map(entries);
		}

		// runs a value through the map
		long getValue(long key) {
			for(Entry entry : entries) {
				long sourceLow = entry.source;
				long sourceHigh = entry.source + entry.range - 1;
				if(sourceLow <= key && key <= sourceHigh) {
					long distance = key - sourceLow;
					return entry.destination + distance;
				}
			}

			return key;
		}

		// runs a value backwards through the map to find the original value
		// uses binary search for the SPEED
		long getKey(long value) {
			int low = 0;
			int high = entries.length - 1;

			while(low <= high) {
				int mid = (low + high) >>> 1;
				Entry entry = entries[mid];
				long destinationLow = entry.destination;
				long destinationHigh = destinationLow + entry.range - 1;

				if(destinationHigh < value) {
					low = mid + 1;
				} else if(destinationLow > value) {
					high = mid - 1;
				} else {
					return entry.source + value - destinationLow;
				}
			}

			return value;
		}

		// finds the range with the highest destination
		PrimitiveIterator.OfLong getMaxRange() {
			long best = 0;
			for(Entry entry : entries) {
				long l = entry.destination + entry.range;
				if(l > best) {
					best = l;
				}
			}
			long highest = best;
			return LongStream.range(0, highest).iterator();
		}

		record Entry(long destination, long source, long range) {
			static Entry create(String line) {
				String[] parts = line.split(" ");
				long destination = Utils.fastParseLong(parts[0]),
						source = Utils.fastParseLong(parts[1]),
						range = Utils.fastParseLong(parts[2]);
				return new Entry(destination, source, range);
			}
		}
	}
}

File: src/2023/kotlin/day5.kt
import Map.Entry
import dev.rdh.aoc.*

private fun PuzzleInput.part1(): Any? {
    val parsedInput = sections.map { it.lines() }
    val seeds = parsedInput[0][0].substring(7).split(" ").map { it.toLong() }
    val maps = parsedInput.drop(1).map {
        Map(it.drop(1).map(Entry::of).sortedBy(Entry::destination).toTypedArray())
    }

    return seeds.minOf { maps.fold(it) { acc, map -> map.getValue(acc) } }
}

private fun PuzzleInput.part2(): Any? {
    val parsedInput = sections.map { it.lines() }
    val seeds = parsedInput[0][0].substring(7).split(" ").map { it.toLong() }
    val maps = parsedInput.drop(1).map {
        Map(it.drop(1).map(Entry::of).sortedBy(Entry::destination).toTypedArray())
    }

    val ranges = seeds.chunked(2).map { it[0] until it[0] + it[1] }

    for (next in maps.last().maxRange()) {
        val changed = maps.foldRight(next) { map, acc -> map.getKey(acc) }
        if (ranges.any { changed in it }) {
            return next
        }
    }

    return "No solution found"
}

private data class Map(val entries: Array<Entry>) {
    data class Entry(val destination: Long, val source: Long, val range: Long) {
        companion object {
            fun of(line: String): Entry {
                val (d, s, r) = line.split(" ").map { it.toLong() }
                return Entry(d, s, r)
            }
        }
    }

    fun getValue(key: Long): Long {
        entries.forEach {
            if(key in it.source until it.source + it.range) {
                return it.destination + (key - it.source)
            }
        }

        return key
    }

    fun getKey(key: Long): Long {
        var low = 0
        var high = entries.size - 1
        while (low <= high) {
            val mid = (low + high) ushr 1
            val midVal = entries[mid]
            when {
                midVal.destination < key -> low = mid + 1
                midVal.destination > key -> high = mid - 1
                else -> return midVal.source
            }
        }

        return key
    }

    fun maxRange() = 0..entries.maxOf { it.range }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Map

        return entries.contentEquals(other.entries)
    }

    override fun hashCode(): Int {
        return entries.contentHashCode()
    }
}

fun main() {
    val input = getInput(2023, 5)

    var start = System.nanoTime()
    var result = input.part1()
    var end = System.nanoTime()
    println("--- Part 1: %.2fms ---".format((end - start) / 1e6))
    println(result)

    start = System.nanoTime()
    result = input.part2()
    end = System.nanoTime()
    println("--- Part 2: %.2fms ---".format((end - start) / 1e6))
    println(result)
    println("----------------------")
}